src/shogun/lib/basetag.h

```diff
@@ -0,0 +1,100 @@
+/*
+ * -*- coding: utf-8 -*-
+ * vim: set fileencoding=utf-8
+ *
+ * Copyright (c) 2016, Shogun-Toolbox e.V. <shogun-team@shogun-toolbox.org>
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ *  1. Redistributions of source code must retain the above copyright notice,
+ *     this list of conditions and the following disclaimer.
+ *
+ *  2. Redistributions in binary form must reproduce the above copyright
+ *     notice, this list of conditions and the following disclaimer in the
+ *     documentation and/or other materials provided with the distribution.
+ *
+ *  3. Neither the name of the copyright holder nor the names of its
+ *     contributors may be used to endorse or promote products derived from
+ *     this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ * Authors: 2016, Sergey Lisitsyn <lisitsyn.s.o@gmail.com>
+ *          2016, Sanuj Sharma <sanuj.sharma.in@gmail.com>
+ */
+
+#ifndef _BASETAG_H_
+#define _BASETAG_H_
+
+#include <string>
+
+namespace shogun
+{
+    class BaseTag
+    {
+    public:
+
+        BaseTag(const std::string& _name) : 
+            name_(_name), hash_(std::hash<std::string>()(_name))
+        {
+        }
+
+        BaseTag(const BaseTag& other) : 
+            name_(other.name_), hash_(other.hash_)
+        {
+        }
+
+        BaseTag& operator=(const BaseTag& other)
+        {
+            name_ = other.name_;
+            hash_ = other.hash_;
+            return *this;
+        }
+        
+        inline std::string name() const
+        {
+            return name_;
+        }
+
+        inline std::size_t hash() const
+        {
+            return hash_;
+        }
+
+        bool operator==(const BaseTag &other) const
```

Comment:
Maybe it is better to implement this as friend - `inline bool operator==(const BaseTag& first, const BaseTag& second)`.